### PR TITLE
cleanup phpdoc and BatchResult interface

### PR DESCRIPTION
**Pull request by dbu**
mainly adjusting the exceptions, but i also went over the BatchResult interface. we had duplicated methods and i reordered them to be consistent.

isSuccessful and hasResponseFor are the same
as are isFailed and hasExceptionFor

i went with isSuccessful and isFailed, but maybe hasExceptionFor / hasResponseFor would be more straightforward. wdyt?

### CODE REVIEW

File: src/Exception/TransferException.php [hunk 1]
```diff
@@ -9,6 +9,6 @@
  *
  * @author Márk Sági-Kazár <mark.sagikazar@gmail.com>
  */
-class TransferException extends \RuntimeException implements Exception
+abstract class TransferException extends \RuntimeException implements Exception
```
Comment (dbu) [inline src/Exception/TransferException.php:12]: i think we never need to directly instantiate this exception. or do we?

File: src/Exception/RequestException.php [hunk 2]
```diff
@@ -37,19 +40,4 @@ public function getRequest()
...
         return $this->request;
     }
-
-    /**
-     * @param RequestInterface $request
-     * @param \Exception       $e
-     *
-     * @return RequestException
-     */
-    public static function wrapException(RequestInterface $request, \Exception $e)
```
Comment (dbu) [inline src/Exception/RequestException.php:47]: not sure what this was for. i guess we should go for the correct exception and just use new. its also weird how the passed request would get lost if $e is already a RequestException...

File: src/Exception/TransferException.php (hunk 1 above)
Comment (sagikazarmark) [inline src/Exception/TransferException.php:12]: Don't think so.

**Comment by sagikazarmark** [comment]
I think isSuccessful and isFailed are better from a domain point of view. We are curious about if the request was successful or failed after all.

**Comment by dbu** [comment]
BatchException never neds to have a previous exception, right? it currently does not have that option, but it also would not make much sense.

File: src/Exception/RequestException.php (hunk 2 above)
Comment (sagikazarmark) [inline src/Exception/RequestException.php:47]: It's from guzzle, free to remove it.

**Comment by sagikazarmark** [comment]
Agreed. Actually there are several previous exceptions which are exposed through a custom API.

File: spec/Exception/TransferExceptionSpec.php [hunk 3]
```diff
@@ -1,24 +0,0 @@
-<?php
-
-namespace spec\Http\Client\Exception;
-
-use PhpSpec\ObjectBehavior;
-use Prophecy\Argument;
-
-class TransferExceptionSpec extends ObjectBehavior
```
Comment (dbu) [inline spec/Exception/TransferExceptionSpec.php:8]: or is there a way to spec these things without instantiating the class?

**Comment by dbu** [comment]
this looks ready from my point, please merge if you agree. i added https://github.com/php-http/utils/pull/12 to adjust the BatchResult implementation.

File: spec/Exception/TransferExceptionSpec.php (hunk 3 above)
Comment (sagikazarmark) [inline spec/Exception/TransferExceptionSpec.php:8]: You can add a stub, like in case of BatchRequest trait and test that.

**Comment by sagikazarmark** [comment]
Do you want to add tests for TransferException prior to merging?

**Comment by dbu** [comment]
added the TransferExcpetion spec back. ready to merge i hope?

**Comment by sagikazarmark** [comment]
Go ahead once tests are finished.

**Pull request merged**